src/main/java/edu/cornell/kfs/module/purap/document/service/impl/CuB2BShoppingServiceImpl.java

```diff
@@ -217,47 +225,83 @@ public void setUserFavoriteAccountService(
 	}
 
+    public void setJaggaerRoleService(JaggaerRoleService jaggaerRoleService) {
+        this.jaggaerRoleService = jaggaerRoleService;
+    }
+
+    @Override
+    public void setB2bDao(B2BDao b2bDao) {
+        super.setB2bDao(b2bDao);
+        this.b2bDao = b2bDao;
+    }
+
+    @Override
     public void setBusinessObjectService(BusinessObjectService businessObjectService) {
         super.setBusinessObjectService(businessObjectService);
         this.businessObjectService = businessObjectService;
     }
 
+    @Override
     public void setDocumentService(DocumentService documentService) {
         super.setDocumentService(documentService);
         this.documentService = documentService;
     }
 
+    @Override
     public void setParameterService(ParameterService parameterService) {
         super.setParameterService(parameterService);
         this.parameterService = parameterService;
     }
 
+    @Override
     public void setPersistenceService(PersistenceService persistenceService) {
         super.setPersistenceService(persistenceService);
         this.persistenceService = persistenceService;
     }
 
+    @Override
     public void setPhoneNumberService(PhoneNumberService phoneNumberService) {
         super.setPhoneNumberService(phoneNumberService);
         this.phoneNumberService = phoneNumberService;
     }
 
+    @Override
     public void setPurchasingService(PurchasingService purchasingService) {
         super.setPurchasingService(purchasingService);
         this.purchasingService = purchasingService;
     }
 
+    @Override
     public void setPurapService(PurapService purapService) {
         super.setPurapService(purapService);
         this.purapService = (CuPurapService) purapService;
     }
 
+    @Override
     public void setVendorService(VendorService vendorService) {
         super.setVendorService(vendorService);
         this.vendorService = vendorService;
     }
 
+    @Override
+    public String getPunchOutUrl(Person user, JaggaerRoleSet roleSet) {
+        List<String> roles = jaggaerRoleService.getJaggaerRoles(user, roleSet);
+        B2BInformation b2b = getB2bShoppingConfigurationInformation();
+        String cxml = getPunchOutSetupRequestMessage(user, b2b, roles);
+        String response = b2bDao.sendPunchOutRequest(cxml, b2b.getPunchoutURL());
+        PunchOutSetupResponse posr = B2BParserHelper.getInstance().parsePunchOutSetupResponse(response);
+        return posr.getPunchOutUrl();
+    }
+
+    @Override
     public String getPunchOutSetupRequestMessage(Person user, B2BInformation b2bInformation) {
+        String preAuthRole = getPreAuthValue(user.getPrincipalId());
+        String viewRole = getViewValue(user.getPrincipalId());
+        List<String> roles = List.of(preAuthRole, viewRole);
+        return getPunchOutSetupRequestMessage(user, b2bInformation, roles);
+    }
+
+    protected String getPunchOutSetupRequestMessage(Person user, B2BInformation b2bInformation, List<String> roles) {
```

Comment:
instead of taking in List<String> roles, it could take in JaggaerRoleSet roleSet

Reply:
I'll adjust this to accept a role set instead.